File: 187.repeated-dna-sequences.js
/*
 * @lc app=leetcode id=187 lang=javascript
 *
 * [187] Repeated DNA Sequences
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string[]}
 */
var findRepeatedDnaSequences = function (s) {
  const res = [];
  const lookup = new Set();
  for (let i = 0; i + 9 < s.length; ++i) {
    const t = s.slice(i, i + 10);
    if (!lookup.has(t)) {
      lookup.add(t);
    } else if (!res.includes(t)) {
      res.push(t);
    }
  }
  return res;
};
// @lc code=end

File: 187.repeated-dna-sequences.py
#
# @lc app=leetcode id=187 lang=python3
#
# [187] Repeated DNA Sequences
#

# Time:  O(n)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        res, lookup = set(), set()
        if len(s) < 10:
            return []
        for i in range(len(s) - 9):
            j = i + 10
            if s[i:j] in lookup:
                res.add(s[i:j])
            else:
                lookup.add(s[i:j])
        return list(res)


# @lc code=end
